### PR TITLE
Optimize Query::get_component

### PR DESCRIPTION
**Pull request by TheRawMeatball**
Right now Query::get_component creates a new EntityRef per get call, and then uses get_unchecked methods to get the data, causing more lookups and branches than necessary. This PR optimizes get_component by taking advantage of data already cached on Query s and taking advantage of derive component to const-ify certain branches.

### CODE REVIEW

File: crates/bevy_ecs/src/system/query.rs [hunk 1]
```diff
@@ -772,29 +789,54 @@ where
...
-        let world = self.world;
-        let entity_ref = world
-            .get_entity(entity)
+        let (component_id, write_flag) = self
+            .state
+            .fetch_state
+            .get_id::<T>()
+            .ok_or(QueryComponentError::MissingWriteAccess)?;
+
+        if write_flag != RWAccess::Write {
```
Comment (BoxyUwU) [inline crates/bevy_ecs/src/system/query.rs:798]: Can we add a comment or something that its ok that `get_id` returns immediately on finding `T` as if there is a `RWAccess::Write` there wont be any other query params for T. e.g. a query like `Query<(&T, &mut T)>` cant exist so we dont care that `get_id` would return `RwAccess::Read`

File: crates/bevy_ecs/src/query/filter.rs [hunk 2]
```diff
@@ -503,6 +518,10 @@ macro_rules! impl_tick_filter {
             fn matches_table(&self, table: &Table) -> bool {
                 table.has_column(self.component_id)
             }
+
+            fn get_id<Comp: 'static>(&self) -> Option<(ComponentId, RWAccess)> {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/filter.rs:522]: Can we make these functions const?

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
> looks good other than this

I'm counting this as an approval.

**Comment by cart** [comment]
I'd like to see some benchmark results before merging this, given that performance is the motivator behind the change, this increases code complexity, and bevy_ecs perf is notoriously fickle.

**Comment by alice-i-cecile** [comment]
@TheRawMeatball, when you get a chance can you rebase this? We have ECS benchmarks now; I can give running the comparison a shot.

**Comment by TheRawMeatball** [comment]
I'll rebase this and run benches later today.

**Comment by alice-i-cecile** [comment]
@TheRawMeatball merge conflicts and failing CI for you <3 We should also double-check those benchmarks, if only for the release notes hype.

**Comment by tygyh** [comment]
Since get_component has changed a lot since this PR was opened, is there a reason to keep it open?

**Pull request closed without merging**